temporalio/worker/_workflow_instance.py

```diff
@@ -1415,14 +1439,47 @@ def workflow_upsert_search_attributes(
                         else [update.value]
                     )
 
+    async def workflow_sleep(
+        self, duration: float, *, summary: Optional[str] = None
+    ) -> None:
+        user_metadata = (
+            temporalio.api.sdk.v1.UserMetadata(
+                summary=self._payload_converter.to_payload(summary)
+            )
+            if summary
+            else None
+        )
+        self._timer_impl(
+            duration, lambda _: None, options=_TimerOptions(user_metadata=user_metadata)
+        )
+
     async def workflow_wait_condition(
-        self, fn: Callable[[], bool], *, timeout: Optional[float] = None
+        self,
+        fn: Callable[[], bool],
+        *,
+        timeout: Optional[float] = None,
+        timeout_summary: Optional[str] = None,
     ) -> None:
         self._assert_not_read_only("wait condition")
         fut = self.create_future()
         self._conditions.append((fn, fut))
+        user_metadata = (
+            temporalio.api.sdk.v1.UserMetadata(
+                summary=self._payload_converter.to_payload(timeout_summary)
+            )
+            if timeout_summary
+            else None
+        )
+        self._next_timer_options = _TimerOptions(user_metadata=user_metadata)
```

Comment:
I am not sure `call_later` is invoked if `timeout` is `None`, so this may leak to other timers. Also, I am not sure you can guarantee this timer options will apply to the immediate `call_later` if many wait conditions are called concurrently. I wonder if you can use `contextvars` here and unset in a `finally`. Or maybe even better, reading the `Timeout.reschedule` code inside of CPython `asyncio` source, I wonder if you can do `setattr` on the `asyncio.current_task` (using `getattr` later) and unset in `finally`. May be worth trying that last option there with a test, that will ensure the least amount of pollution.

(I regret using `asyncio.wait_for` here for multiple reasons including this and the fact that they changed the impl on us in newer Python versions)

Reply:
I can't even guarantee `call_later` is called by `wait_for` (I think `timeout` as `None` won't call it for instance). We need to try our best to localize the next timer options to _this_ call, which I am surprised `contextvars` does not do, but maybe mutating the current task will do it.